File: trees_and_graphs/reorder_routes_to_make_all_paths_lead_to_the_city_zero.cpp
#include <gtest/gtest.h>

#include <set>
#include <unordered_map>
#include <vector>

using Graph_t = std::unordered_map<int, std::vector<int>>;

//! @brief DFS starting at node for determining number of edges to swap
//! @param[in]     node  Starting node to travel from
//! @param[in]     graph Undirected mapping of node to its neighbors
//! @param[in]     roads Set of original directed edges between nodes
//! @param[in,out] seen  Vector of nodes that have been visited
//! @return Number of edges to swap to make all paths lead to node
static int dfsRecursive(int                              node,
                        const Graph_t&                   graph,
                        const std::set<std:vector<int>>& roads,
                        std::vector<bool>&               seen)
{
    int ans {};

    for (const int neighbor : graph.at(node))
    {
        if (not seen[neighbor])
        {
            if (roads.find({node, neighbor}) != roads.end())
            {
                ++ans;
            }

            seen[neighbor] = true;
            ans += dfsRecursive(neighbor);
        }
    }

    return ans;
}

//! @brief Get min number of swaps to make all paths lead to the city zero
//! @param[in] n           Number of cities numbered from 0 to n - 1
//! @param[in] connections Roads between city x and city y
//! @return Minimum number of swaps needed so every city can reach city 0
static int minReorder(int n, const std::vector<std::vector<int>>& connections)
{
    //! @details Time complexity is O(N) because only visit each node once and
    //!          do constant work.
    //!          Space complexity is O(N) since roads, graph, and seen all use
    //!          up to O(N) space at most

    std::vector<bool>          seen(n, false);
    std::set<std::vector<int>> roads {};
    Graph_t                    graph {};

    for (const auto& edge : connections)
    {
        const int x {edge[0]};
        const int y {edge[1]};

        graph[x].push_back(y);
        graph[y].push_back(x);
        roads.emplace({x, y});
    }

    seen[0] = true;
    return dfsRecursive(0, graph, roads, seen);
}

File: trees_and_graphs/reorder_routes_to_make_all_paths_lead_to_the_city_zero.js
/**
 * @param {number} n
 * @param {number[][]} connections
 * @returns {number}
 */
var minReorder = function(n, connections) {
    const convertToHash = (row, col) => {
        return row + "," + col;
    }

    const dfsRecursive = node => {
        let ans = 0;
        for (const neighbor of graph.get(node))
        {
            if (!seen[neighbor])
            {
                if (roads.has(convertToHash(node, neighbor)))
                {
                    ans++;
                }

                seen[neighbor] = true;
                ans += dfsRecursive(neighbor);
            }
        }

        return ans;
    }

    const dfsIterative = start => {
        let ans = 0;

        let stack = [start];
        while (stack.length)
        {
            const node = stack.pop();
            
            for (const neighbor of graph.get(node))
            {
                if (!seen[neighbor])
                {
                    if (roads.has(convertToHash(node, neighbor)))
                    {
                        ans++;
                    }

                    seen[neighbor] = true;
                    stack.push(neighbor);
                }
            }
        }

        return ans;
    }

    let roads = new Set();
    let graph = new Map();
    let seen = new Array(n).fill(false);

    for (let i = 0; i < n; i++)
    {
        graph.set(i, []);
    }

    for (const [x, y] of connections)
    {
        graph.get(x).push(y);
        graph.get(y).push(x);
        roads.add(convertToHash(x, y));
    }

    seen[0] = true;
    return dfsRecursive(0);
    // return dfsIterative(0);
};